nexus/tests/integration_tests/instances.rs

```diff
@@ -6028,6 +6027,222 @@ async fn test_instance_v2p_mappings(cptestctx: &ControlPlaneTestContext) {
     }
 }
 
+#[nexus_test]
+async fn test_instance_force_terminate(cptestctx: &ControlPlaneTestContext) {
+    let client = &cptestctx.external_client;
+    let apictx = &cptestctx.server.server_context();
+    let nexus = &apictx.nexus;
+    let instance_name = "test-instance-please-ignore";
+    let already_gone_name = "what-instance";
+    make_forgotten_instance(
+        &cptestctx,
+        already_gone_name,
+        InstanceAutoRestartPolicy::Never,
+    )
+    .await;
+
+    // Create an instance.
+    let instance = create_instance(client, PROJECT_NAME, instance_name).await;
+    let instance_id = InstanceUuid::from_untyped_uuid(instance.identity.id);
+    instance_simulate(nexus, &instance_id).await;
+
+    let instance = dbg!(
+        instance_post(&client, &instance_name, InstanceOp::ForceTerminate)
+            .await
+    );
+    // The instance force terminate endpoint will attempt to run the
+    // instance-update saga to completion, so the instance should go to Stopped
+    // immediately from the caller's perspective.
+    assert_eq!(instance.runtime.run_state, InstanceState::Stopped);
+
+    // A subsequent force-terminate should be okay because we're already
+    // terminated.
+    let instance = dbg!(
+        instance_post(&client, &instance_name, InstanceOp::ForceTerminate)
+            .await
+    );
+    assert_eq!(instance.runtime.run_state, InstanceState::Stopped);
+
+    // Okay, what if the instance is already gone?
+    let instance = dbg!(
+        instance_post(&client, &already_gone_name, InstanceOp::ForceTerminate)
+            .await
+    );
+    // This time, the instance will go to `Failed` rather than `Stopped` since
+    // sled-agent is no longer aware of it.
+    assert_eq!(instance.runtime.run_state, InstanceState::Failed);
```

Comment:
fwiw this initially seemed concerning ("if you force-stop an instance in the right state, now you have a failed instance?"), but i think i've figured out the preconditions by reading more carefully: this is specifically if you force-stop an instance where Nexus believes a sled-agent is responsible for the instance, but that sled-agent doesn't know it exists, and that _already_ is a sign that something has gone wrong. so it's an instance that probably should be failed anyway, we just hadn't figured that out yet?

if it helps on wording, i think "already gone" is what primed me to think this might be applicable in some kind of race around stopping.

that said, i _think_ there's a race where if  you force-stop an instance alongside a call to `instance_stop` you can end up with an instance in Failed where auto-restart would just turn it back on again? specifically if `instance_stop` has reached out to sled-agent, sled-agent responds that it has stopped the instance, but we haven't actually recorded that state update yet, then `force_stop` could show up and see Nexus thinks that sled-agent has the VMM while that sled-agent thinks it's gone.

i haven't looked at the `instance_stop` code much yet so maybe i'm misunderstanding and there's not actually something there..? but if i'm following these bits right, it seems kind of unfortunate that trying too enthusiastically to stop an instance could end up making it auto-restart.

Reply:
Multiple racing stop attempts shouldn't transition an instance to failed, since the query in `mark_vmm_failed` that transitions the VMM to failed is conditional on the sequence number of the VMM record remaining the same as when the VMM was read. If the VMM was destroyed between when we read the record initially and when we asked the sled-agent to destroy it, the sequence number will have advanced, and we won't move it to failed. It only moves to failed if it was already nonexistent.

I see your point about auto-restarts, though: perhaps we should just always move a force-stopped instance to `Stopped` when encountering a sled-agent error that would otherwise move it to `Failed`, since you're trying to stop it anyway, which presumably means you have other plans for the instance? I dunno. @gjcolombo, what do you think?

Reply:
It occurs to me that we should probably also bail out early if we see the VMM record is already `Destroyed`, though, rather than asking the sled-agent to destroy it again...

Reply:
ok! i see your point about the non-raciness now - if there's still a Propolis ID for an instance, the `force-terminate` call might cause Nexus to get to `instance_ensure_unregistered` and in fact ask `sled-agent` to ensure the Propolis has been terminated, but in that case `sled-agent`'s `ensure_unregistered` would return `VmmUnregisterResponse { updated_runtime: None }` rather than an error that the Propolis that should be unregistered is already gone. to the point of sled-agent docs, `ensure_unregistered` is dempotent.

so, among other things, new `Err(InstanceStateChangeError::SledAgent(e)) if e.vmm_gone()` arm in `instance.rs` _shouldn't_ ever be reached, and _that_ case of marking an instance failed doesn't matter so much.

fully convinced it's not racy now, thanks for bearing with me.